### PR TITLE
SNOW-567560 Add split view to Snowpark Python API reference

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-567560

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

Removed templates that are unused.
Used `autosummary` on functions/exception/attributes/classes directly instead of `automodule` to generate separate indexed pages. This creates manual work, but `automodule` does not generate these indexes correctly, which seems to be [an open feature request](https://github.com/sphinx-doc/sphinx/issues/7912) for sphinx-autosummary.
On top of that, the team agree we should organize our API's by functionality instead of by module structure, which will also remove most (if not all) usage of `automodule`.

### CODE REVIEW

**Comment by sfc-gh-stan** [comment]
The warning in `Build Doc` is caused by [an open issue](https://github.com/sphinx-doc/sphinx/issues/9884) of autosummary. My workaround is to define `is_cached` as a property, we can revert this once the issue becomes resolved.

File: docs/source/conf.py [hunk 1]
```diff
@@ -95,3 +95,5 @@
 
 # Disable footer message "Built with Sphinx using a theme provided by Read the Docs."
 html_show_sphinx = False
+
+add_module_names = False
```
Comment (sfc-gh-stan) [inline docs/source/conf.py:99]: Will fix.

File: docs/source/functions.rst [hunk 2]
```diff
@@ -0,0 +1,430 @@
+snowflake.snowpark.functions
+================================
+Provides utility and SQL functions that generate :class:`~snowflake.snowpark.Column` expressions that you can pass to :class:`~snowflake.snowpark.DataFrame` transformation methods.
```
Comment (sfc-gh-stan) [inline docs/source/functions.rst:3]: When using `automodule`, this docstring will be extracted automatically from `src/snowflake/snowpark/functions.py`, I have yet to figure out how to do that using `currentmodule`.

File: docs/source/index.rst [hunk 3]
```diff
@@ -10,20 +10,14 @@ Snowpark simplifies the process of building complex data pipelines and allows yo
...
-   snowflake.snowpark.udf
-   snowflake.snowpark.udtf
-   snowflake.snowpark.exceptions
-
-:ref:`genindex`
-
+.. toctree::
+   :maxdepth: 2
+
+   snowpark_classes.rst
```
Comment (sfc-gh-stan) [inline docs/source/index.rst:16]: Will fix.

File: src/snowflake/snowpark/dataframe.py [hunk 4]
```diff
@@ -520,6 +520,15 @@ def __init__(
     def stat(self) -> DataFrameStatFunctions:
         return self._stat
 
+    @property
+    def is_cached(self) -> bool:
+        """Whether the dataframe is cached"""
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/dataframe.py:525]: Add period .

File: docs/source/conf.py [hunk 5]
```diff
@@ -43,20 +43,21 @@
...
 autodoc_default_options = {
     "members": True,  # Include all methods.
     "member-order": "alphabetical",  # 'alphabetical', by member type ('groupwise') or source order (value 'bysource')
     "undoc-members": True,  # If set, autodoc will also generate document for the members not having docstrings
-    "private-members": False,  # don't generate document for the private members (like _private or __private)
-    # 'special-members': '',  # for example, '__init__'
-    "inherited-members": False,
+    # "private-members": '',  # don't generate document for the private members (like _private or __private)
+    # "special-members": '',  # for example, '__init__'
+    # "inherited-members": '',
```
Comment (sfc-gh-sfan) [inline docs/source/conf.py:55]: I'm a bit confused about the changes here. Why do we comment certain settings?

Comment (sfc-gh-stan) [reply docs/source/conf.py:55]: This is because specifying an empty string or `None` is same as not specifying. This is a good point, I should probably just remove the comments to avoid confusion.

File: docs/source/exceptions.rst [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+This package contains all Snowpark client-side exceptions.
+
+.. currentmodule:: snowflake.snowpark.exceptions
+
+
+   
+.. rubric:: Exceptions
+
+.. autosummary::
+    :toctree: api/
```
Comment (sfc-gh-stan) [inline docs/source/exceptions.rst:12]: Note: this will change the URL `https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/_autosummary/snowflake.snowpark.html` by replacing `_autosummary` with `api`. I think this way makes more sense, but if it's considered breaking I can revert this change.

File: docs/source/conf.py [hunk 7]
```diff
@@ -43,20 +43,21 @@
...
     "show-inheritance": True,
-    # 'exclude-members': ''   # for example,
+    # "exclude-members": ''#
```
Comment (sfc-gh-stan) [inline docs/source/conf.py:57]: Remove `#` here.

File: docs/source/exceptions.rst (hunk 6 above)
Comment (sfc-gh-sfan) [reply docs/source/exceptions.rst:12]: I prefer the new name, but I'm not very sure about the impact of URL change. I wonder if it is possible to add redirect to the original link? 🤔

File: docs/source/index.rst [hunk 8]
```diff
@@ -10,20 +10,14 @@ Snowpark simplifies the process of building complex data pipelines and allows yo
...
-   snowflake.snowpark
-   snowflake.snowpark.functions
-   snowflake.snowpark.table_function
-   snowflake.snowpark.stored_procedure
-   snowflake.snowpark.types
-   snowflake.snowpark.udf
-   snowflake.snowpark.udtf
-   snowflake.snowpark.exceptions
-
-:ref:`genindex`
```
Comment (sfc-gh-mabrennan) [inline docs/source/index.rst:28]: :ref:`genindex` renders as a [link to the index](https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/genindex.html). It's pretty handy so I hope we can keep it. Here's a screenshot:
image

Comment (sfc-gh-mabrennan) [reply docs/source/index.rst:28]: It can go somewhere else if you prefer.

**Review by sfc-gh-mabrennan — APPROVED** [review]
Very cool stuff. Thanks for making these improvements!

I suggest updating the end of the snowpark-python/docs/README.md file which explains how to build the API docs as it mentions some templates that got changed.

File: docs/source/exceptions.rst (hunk 6 above)
Comment (sfc-gh-mabrennan) [reply docs/source/exceptions.rst:12]: What benefit does changing this directory name from "_autosummary" to "api" give us? Is breaking all the links and messing up peoples' bookmarks worth it? I don't know how to add a redirect in this situation.

File: src/snowflake/snowpark/dataframe.py [hunk 9]
```diff
@@ -520,6 +520,15 @@ def __init__(
...
 
+    @property
+    def is_cached(self) -> bool:
+        """Whether the dataframe is cached."""
+        return self._is_cached
+
+    @is_cached.setter
+    def is_cached(self, value: bool):
+        self._is_cached = value
+
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/dataframe.py:531]: Is this related to split view?

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/dataframe.py:531]: Yes, this is a workaround for https://github.com/sphinx-doc/sphinx/issues/9884 of autosummary.

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/dataframe.py:531]: Ok. The change looks good to me. Only concern is the breaking links which Mary mentioned above. Do you have a plan to deal with those?

**Review by sfc-gh-aalam — APPROVED** [review]

File: docs/source/exceptions.rst (hunk 6 above)
Comment (sfc-gh-jdu) [reply docs/source/exceptions.rst:12]: Is it possible that we keep `_autosummary` here, but create a new page with `api`, and redirect it to `_autosummary`?

File: docs/source/functions.rst (hunk 2 above)
Comment (sfc-gh-jdu) [reply docs/source/functions.rst:3]: So you mean text in this file has to be manually extracted from src/snowflake/snowpark/functions.py? Hmm it might not be a good idea since it's hard to keep them the same if someone has a change in src/snowflake/snowpark/functions.py?

File: docs/source/types.rst [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+.. currentmodule:: snowflake.snowpark.types
+
+   
+
+.. rubric:: Module Attributes
+
+.. autosummary::
+    :toctree: api/
+
+    Variant
```
Comment (sfc-gh-jdu) [inline docs/source/types.rst:14]: Could you move these two to the following section as well?

File: src/snowflake/snowpark/dataframe.py [hunk 11]
```diff
@@ -520,6 +520,15 @@ def __init__(
     def stat(self) -> DataFrameStatFunctions:
         return self._stat
 
+    @property
+    def is_cached(self) -> bool:
+        """Whether the dataframe is cached."""
+        return self._is_cached
+
+    @is_cached.setter
+    def is_cached(self, value: bool):
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/dataframe.py:529]: Probably we shouldn't expose setter since it's not supposed to be set by users

File: docs/source/functions.rst (hunk 2 above)
Comment (sfc-gh-stan) [reply docs/source/functions.rst:3]: Yes, I don't have a better workaround for now. What if we just move the text to be in the documentation only?

File: docs/source/exceptions.rst (hunk 6 above)
Comment (sfc-gh-stan) [reply docs/source/exceptions.rst:12]: I think redirecting is more related to how the webpages are hosted, `sphinx` will complain if it detects duplicate page indexes/paths. I'll reach out to Mary to learn about the hosting configs.
CC: @sfc-gh-aalam

File: docs/source/types.rst (hunk 10 above)
Comment (sfc-gh-stan) [reply docs/source/types.rst:14]: I think these two refer to the type hint declaration: 
```
#: The type hint for annotating Variant data when registering UDFs.
Variant = TypeVar("Variant")

#: The type hint for annotating Geography data when registering UDFs.
Geography = TypeVar("Geography")
```
 In the [current API reference](https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/_autosummary/snowflake.snowpark.types.html), these two are identified as module attributes as well.

File: docs/source/functions.rst (hunk 2 above)
Comment (sfc-gh-stan) [reply docs/source/functions.rst:3]: TODO: Let's have a JIRA for this.

**Review by sfc-gh-jdu — APPROVED** [review]
